Make FPGA tests to pass on CI targets (SPI, analogIn, PWM)

mprse opened the pull request:
### Description

I run the FPGA tests on potential CI targets and got the following results:

  | GPIO | GPIO_IRQ | AnalogIn | PWM | I2C | SPI | UART
-- | -- | -- | -- | -- | -- | -- | --
NUCLEO_F429ZI | OK | OK | OK | OK | OK | OK | OK
K64F | OK | OK | OK | OK | OK | OK | OK
NUCLEO_F411RE | OK | OK | **FAIL: D13 pin not supported** | OK | **FAIL** | **TIMEOUT: SPI async issue(lp-ticker ISR execution is too long)** | OK
DISCO_L475VG_IOT01A | **FAIL** | OK | OK | OK | OK | OK | **FAIL**
NRF52840_DK | OK | **FAIL** | OK | **FAIL: Driver bug/hw limitation** | **FAIL** | **TIMEOUT: 16 bit symbol not supported** | **TIMEOUT**
NUCLEO_F303RE | OK | OK | **FAIL: D13 pin not supported** | OK | OK | OK | **TIMEOUT**
LPC55S69_NS | **FAIL** | **FAIL** | OK | NOT SUPPORTED | **FAIL** | OK | **TIMEOUT**
NUCLEO_L073RZ | OK | OK | **FAIL: D13 pin not supported** | OK | OK | OK | OK

This PR should solve the issues with `SPI`, `AnalogIn`, `PWM` tests.

@jeromecoutant @LMESTM 
I noticed the problem with the SPI async mode on `NUCLEO_F411RE`. It looks like the system tick breaks the transmission. Can you take a look? More details in the commit description.
Also, it looks like `PA_5` pin can not be used as `analogin`. I removed it from the pinmap table. Is this an acceptable solution?

### Pull request type

    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

@maciejbocianski @fkjagodzinski 
@0xc0170 @jamesbeyond

mprse committed:
NUCLEO_F411RE, NUCLEO_L073RZ, NUCLEO_F303RE: Disable Analogin D13(PA_5) pin.

Analogin test fails on D13(PA_5) pin. When logic one (3.3V) is provided on this pin ADC reads 0.86 value. On other pins we got 0.98.
This is caused because this pin is connected to led2.

mprse committed:
SPI fpga test: disable 16 bit test case for NRF52840_DK (not supported)

mprse committed:
NRF52: Fix PWM driver

Fix pwm nrf

mprse committed:
Update tests-mbed_hal_fpga_ci_test_shield-pwm test (adapt for NRF_52)

Looks like the PWM works fine on NRF52840_DK target, but this target has hardware limitation and the max PWM period is ‭32 767‬ us.
That is why test cases when the period is set 50 ms are failing.
The test will be modified to test the 30 ms period instead of 50 ms, so the test can pass on all CI targets.

jeromecoutant commented on targets/targets.json:
```diff
@@ -2622,6 +2622,9 @@
             "MPU"
         ],
+        "device_has_remove": [
+            "SPI_ASYNCH"
```
We don't see any reason to remove SPI_ASYNC for this specific target...

mprse replied:
The main reason is that async mode does not work on this target (test case fails). 
The best option would be to find why. I will try to do further investigation, but I wanted first to check if maybe you know something about this.

jeromecoutant replied:
@ABOSTM

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32F3/TARGET_STM32F303xE/TARGET_NUCLEO_F303RE/PeripheralPins.c:
```diff
@@ -58,7 +58,7 @@ MBED_WEAK const PinMap PinMap_ADC[] = {
 //  {PA_3,       ADC_1, STM_PIN_DATA_EXT(STM_MODE_ANALOG, GPIO_NOPULL, 0, 4, 0)}, // ADC1_IN4 // Connected to STDIO_UART_RX
     {PA_4,       ADC_2, STM_PIN_DATA_EXT(STM_MODE_ANALOG, GPIO_NOPULL, 0, 1, 0)}, // ADC2_IN1
-    {PA_5,       ADC_2, STM_PIN_DATA_EXT(STM_MODE_ANALOG, GPIO_NOPULL, 0, 2, 0)}, // ADC2_IN2 // Connected to LD2 [Green Led]
+//    {PA_5,       ADC_2, STM_PIN_DATA_EXT(STM_MODE_ANALOG, GPIO_NOPULL, 0, 2, 0)}, // ADC2_IN2 // Connected to LD2 [Green Led]
```
Maybe pin should be added to the restricted list ?
To be tested

mprse replied:
I was thinking about this. But it looks like this pin can not be used as Analogin pin (the results are incorrect), so it should be removed from the pinmap table.

jeromecoutant reviewed (changes requested):
Maybe you should separate TESTS update from targets update ?

c1728p9 commented on TESTS/mbed_hal_fpga_ci_test_shield/spi/main.cpp:
```diff
@@ -177,9 +177,9 @@ Case cases[] = {
     Case("SPI - mode testing (MODE_2)", one_peripheral<SPIPort, DefaultFormFactor, spi_test_common<SPITester::Mode2, 8, TRANSFER_SPI_MASTER_WRITE_SYNC, FREQ_1_MHZ> >),
     Case("SPI - mode testing (MODE_3)", one_peripheral<SPIPort, DefaultFormFactor, spi_test_common<SPITester::Mode3, 8, TRANSFER_SPI_MASTER_WRITE_SYNC, FREQ_1_MHZ> >),
-
+#if !defined(TARGET_NRF52840_DK)
```
Why is this test skipped just for the NRF52840? Is this test tied to a defined behavior in the SPI HAL specification? If so users and library may be relying on all targets having this behavior. If this isn't part of the HAL specification then maybe this could be an optional test? What do you think?

mprse replied:
Thanks for the review!
16-bit symbols are not supported on `NRF` target. On master, specification is not very restrictive since it must be consistent with all targets. We also don't have capabilities like on the SPI feature branch, so I can't run this case conditionally. 

I see two options:
- remove the test case permanently, but it would be good to run this test on targets that support 16-bit symbols.
- use the preprocessor to disable the test case on the specific target.

I don't like the first option. Let's consider that we are adding the next target which supports only one clock mode. If we choose this approach, then I would have to remove test cases for other modes. 

Also on the feature branch, we have nice specification and tests, but it looks like there are no plans to bring new SPI into the master in the nearest future 😞 Maybe it would be good to add `spi_get_capabilities()` function on the master.

0xc0170 replied:
> Also on the feature branch, we have nice specification and tests, but it looks like there are no plans to bring new SPI into the master in the nearest future 😞 Maybe it would be good to add spi_get_capabilities() function on the master.

How much work it would be?

mprse replied:
We could add `WEAK` definition for all targets which returns that all SPI capabilities are supported. And then implement a valid version for CI targets and disable unsupported capabilities for each CI target. This way we could also extend the SPI test on the master. I think that this should be easy to implement.

0xc0170 replied:
I've realized just now this has not yet been fixed?

mprse replied:
Not at the moment. This must be discussed. I'll try to write an email about this. The same problem we have already here:
https://github.com/ARMmbed/mbed-os/blob/df79609cc56c063ba292972af77f2b833b0fda35/TESTS/mbed_hal_fpga_ci_test_shield/gpio/main.cpp#L67-L72

0xc0170 commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_NRF52/pwmout_api.c:
```diff
@@ -295,9 +295,9 @@ void pwmout_period_us(pwmout_t *obj, int period)
 {
     DEBUG_PRINTF("pwmout_pulsewidt: %f\r\n", pulse);
-
+    
```
added some spaces?

mprse replied:
Fixed

0xc0170 reviewed (approved):
Just one cosmetic change needed

mprse committed:
Add spi_get_capabilities() function to HAL API

Add also default weak version of spi_get_capabilities() which provides default/most common SPI parameters.
This function can be replaced if a specific target has different capabilities.

mprse committed:
SPI fpga test: use get_capabilities() function to skip test cases for unsupported features

jamesbeyond reviewed (approved):
Looks good, well done !
